enhancements/cloud-controller-manager/openstack-cloud-controller-manager.md

```diff
@@ -0,0 +1,218 @@
+
+### Non-Goals
+
+- [Cinder CSI driver integration](https://github.com/openshift/enhancements/pull/437) is out of scope of this work.
+- Deprecating `hyperkube` binary and switching to standalone binaries for all related components (`kube-apiserver`, `kube-scheduler`, `kube-controller-manager` and others) is also not a goal.
+
+## Proposal
+
+Our main goal is to start using Cloud Controller Manager in OpenShift 4 on OpenStack. So we are going to use the implementation available in upstream as a part of [cloud-provider-openstack](https://github.com/kubernetes/cloud-provider-openstack/tree/master/pkg/cloudprovider) repo.
+
+To maintain the lifecycle of the CCM we want to implement a cluster operator, that will handle all administrative tasks: deploy, restore, upgrade, healthchecks, and so on. Later this operator can be reused to manage CCMs for other platforms (AWS, Azure, GCP and so on).
+
+### Action plan
+
+#### Allow installer to create a secret with CCM configuration
+
+Now we generate the in-tree cloud provider config in the installer's `manifests` [asset](https://github.com/openshift/installer/blob/4aac205a59a9102ddf1ae329b3c4c6f0dff7e330/pkg/asset/manifests/openstack/cloudproviderconfig.go). Since the configuration format is the same, we can reuse the code but generate a secret in the CCM namespace instead.
+
+Actions:
+
+- Rework the `manifest` asset to generate a secret with CCM config.
+
+- Update unit tests.
+
+#### Build OpenStack CCM image by OCP automation
+
+To start using OpenStack CCM in OpenShift we need to build its image and make sure it is a part of the OpenShift release image. The CCM image should be automatically tested before it becomes available.
+The upstream repo provides the Dockerfile, so we can reuse it to complete the task. Upstream image is already available in [Dockerhub](https://hub.docker.com/r/k8scloudprovider/openstack-cloud-controller-manager).
+
+Actions:
+
+- Configure CI operator to build OpenStack CCM image.
+
+CI operator will run containerized and End-to-End tests and also push the resulting image in the OCP Quay account.
+
+#### Test Cloud Controller Manager manually
+
+When all required components are built, we can manually deploy OpenStack CCM and test how it works.
+
+Actions:
+
+- Manually install CCM's [daemonset](https://github.com/kubernetes/cloud-provider-openstack/blob/master/manifests/controller-manager/openstack-cloud-controller-manager-ds.yaml) on a working OCP cluster deployed on OpenStack.
+
+- Update configuration of `kube-apiserver`, `kube-controller-manager` and `kubelet` by replacing `--cloud-provider openstack` with `--cloud-provider external` and removing `--cloud-config` parameters.
+
+**Note:** Example of manual testing: https://asciinema.org/a/303399?speed=2
+
+#### Write Cluster Cloud Controller Manager Operator
+
+The operator should be able to create, configure and manage different CCMs (including one for OpenStack) in OpenShift.
```

Comment:
Are these operations the scope of the unit/CI tests?  Are there more/others?

Reply:
I think the main goal is to reach the feature parity with the in-tree cloud provider. So, if e2e-openstack tests pass, then we are good. 
For sure, unit tests for the operator are required.